model_server/app/function_calling/hallucination_handler.py

```diff
@@ -0,0 +1,240 @@
+        logprobs (list): List of log probabilities for each token.
+        state (str): Current state of the handler.
+        mask (list): List of masks indicating the type of each token.
+        parameter_name_done (bool): Flag indicating if parameter name extraction is done.
+        hallucination (bool): Flag indicating if a hallucination is detected.
+        hallucination_message (str): Message describing the hallucination.
+        parameter_name (list): List of extracted parameter names.
+        function_description (dict): Description of functions and their parameters.
+        token_probs_map (list): List mapping tokens to their entropy and variance of entropy.
+        current_token (str): The current token being processed.
+    """
+
+    def __init__(self):
+        """
+        Initializes the HallucinationStateHandler with default values.
+        """
+        self.tokens = []
+        self.logprobs = []
+        self.state = None
+        self.mask = []
+        self.parameter_name_done = False
+        self.hallucination = False
+        self.hallucination_message = ""
+        self.parameter_name = []
+
+        self.token_probs_map = []
+        self.current_token = None
+
+    def process_function(self, apis):
+        self.apis = apis
+        if self.apis is None:
+            raise ValueError("API descriptions not set.")
+        parameter_names = {}
+        for func in self.apis:
+            func_name = func["name"]
+            parameters = func["parameters"]["properties"]
+            parameter_names[func_name] = list(parameters.keys())
+        self.function_description = parameter_names
+        self.function_properties = {x["name"]: x["parameters"] for x in self.apis}
+
+    def process_token(self):
+        """
+        Processes the current token and updates the state and mask accordingly.
+        Detects hallucinations based on the token type and log probabilities.
+        """
+        content = "".join(self.tokens).replace(" ", "")
+        if self.current_token == "<tool_call>":
+            self.mask.append("t")
+            self.check_logprob()
+
+        # Function name extraction logic
+        if self.state == "function_name":
+            if self.current_token not in const.FUNC_NAME_END_TOKEN:
+                self.mask.append("f")
+            else:
+                self.state = None
+                self.check_function_name()
+
+        if content.endswith(const.FUNC_NAME_START_PATTERN):
+            print("function name entered")
+            self.state = "function_name"
+
+        # Parameter name extraction logic
+        if self.state == "parameter_name" and not content.endswith(
+            const.PARAMETER_NAME_END_TOKENS
+        ):
+            self.mask.append("p")
+        elif self.state == "parameter_name" and content.endswith(
+            const.PARAMETER_NAME_END_TOKENS
+        ):
+            self.state = None
+            self.check_parameter_name()
```

Comment:
if hallucination is detected then break here

Reply:
how to handle when hallucinated will be in a different pr